File: src/analyser/SymbolEntry.java
package analyser;

import instruction.InstructionEntry;
import tokenizer.TokenType;

import java.util.ArrayList;
import java.util.List;

public class SymbolEntry {


    //    种类：简单变量var，函数func，参数param等
    private String Kind;
    //    类型：int，double，string,void,如果是函数则存其返回类型
    private TokenType Type;
    //    值,如果是函数，则存其返回值
    private Object value;
    //如果是函数，则是参数类型
    private ArrayList<TokenType> paramsType = new ArrayList<>();

    //    如果是函数
    private int level;
//    指令集
    private ArrayList<InstructionEntry> instructions = new ArrayList<>();
//    本地变量集
    private ArrayList<String> locVars = new ArrayList<>();
//    函数参数集
    private ArrayList<String> paramVars = new ArrayList<>();



    // 是否常量
    private boolean isConstant;
    // 是否有值
    private boolean isInitialized;

    private boolean isGlobal = false;

    private int stackOffset;

    public SymbolEntry(String kind, TokenType type, int level, boolean isConstant, boolean isInitialized,int stackOffset) {
        Kind = kind;
        Type = type;
        this.level = level;
        this.isConstant = isConstant;
        this.isInitialized = isInitialized;
        this.stackOffset = stackOffset;
    }

    public String getKind() {
        return Kind;
    }

    public void setKind(String kind) {
        Kind = kind;
    }

    public TokenType getType() {
        return Type;
    }

    public void setType(TokenType type) {
        Type = type;
    }

    public Object getValue() {
        return value;
    }

    public void setValue(Object value) {
        this.value = value;
    }


    public int getLevel() {
        return level;
    }

    public void setLevel(int level) {
        this.level = level;
    }
    public ArrayList<InstructionEntry> getInstructions() {
        return instructions;
    }

    public void setInstructions(ArrayList<InstructionEntry> instructions) {
        this.instructions = instructions;
    }

    public ArrayList<String> getLocVars() {
        return locVars;
    }

    public void setLocVars(ArrayList<String> locVars) {
        this.locVars = locVars;
    }

    public ArrayList<String> getParamVars() {
        return paramVars;
    }

    public void setParamVars(ArrayList<String> paramVars) {
        this.paramVars = paramVars;
    }

    public boolean isConstant() {
        return isConstant;
    }

    public void setConstant(boolean constant) {
        isConstant = constant;
    }

    public boolean isInitialized() {
        return isInitialized;
    }

    public void setInitialized(boolean initialized) {
        isInitialized = initialized;
    }

    public boolean isGlobal() {
        return isGlobal;
    }

    public void setGlobal(boolean global) {
        isGlobal = global;
    }

    public int getStackOffset() {
        return stackOffset;
    }

    public void setStackOffset(int stackOffset) {
        this.stackOffset = stackOffset;
    }
}

File: bin/analyser/symbol_entry.dart
import '../instruction/instruction_entry.dart';
import '../tokenizer/token_type.dart';

class SymbolEntry {
  // 种类：简单变量var，函数func，参数param等
  String kind;
  // 类型：int，double，string,void,如果是函数则存其返回类型
  TokenType type;
  // 值,如果是函数，则存其返回值
  Object value;
  // 如果是函数，则是参数类型
  List<TokenType> paramsType = [];

  // 如果是函数
  int level;
  // 指令集
  List<InstructionEntry> instructions = [];
  // 本地变量集
  List<String> locVars = [];
  // 函数参数集
  List<String> paramVars = [];

  // 是否常量
  bool _isConstant;
  // 是否有值
  bool _isInitialized;

  bool _isGlobal = false;

  int stackOffset;

  SymbolEntry(this.kind, this.type, this.level, this._isConstant,
      this._isInitialized, this.stackOffset) {}

  String getKind() {
    return kind;
  }

  void setKind(String kind) {
    this.kind = kind;
  }

  TokenType getType() {
    return type;
  }

  void setType(TokenType type) {
    this.type = type;
  }

  Object getValue() {
    return value;
  }

  void setValue(Object value) {
    this.value = value;
  }

  int getLevel() {
    return level;
  }

  void setLevel(int level) {
    this.level = level;
  }

  List<InstructionEntry> getInstructions() {
    return instructions;
  }

  void setInstructions(List<InstructionEntry> instructions) {
    this.instructions = instructions;
  }

  List<String> getLocVars() {
    return locVars;
  }

  void setLocVars(List<String> locVars) {
    this.locVars = locVars;
  }

  List<String> getParamVars() {
    return paramVars;
  }

  void setParamVars(List<String> paramVars) {
    this.paramVars = paramVars;
  }

  bool getIsConstant() {
    return _isConstant;
  }

  void setConstant(bool constant) {
    _isConstant = constant;
  }

  bool isInitialized() {
    return _isInitialized;
  }

  void setInitialized(bool initialized) {
    _isInitialized = initialized;
  }

  bool isGlobal() {
    return _isGlobal;
  }

  void setGlobal(bool global) {
    _isGlobal = global;
  }

  int getStackOffset() {
    return stackOffset;
  }

  void setStackOffset(int stackOffset) {
    this.stackOffset = stackOffset;
  }

  @override
  String toString() {
    return "{ kind: " + kind + "--- type:" + type.toString() + "}";
  }
}
